frontend/catalyst/utils/jax_extras.py

```diff
@@ -383,3 +401,80 @@ def new_inner_tracer(trace: DynamicJaxprTrace, aval) -> DynamicJaxprTracer:
     trace.frame.tracers.append(dt)
     trace.frame.tracer_to_var[id(dt)] = trace.frame.newvar(aval)
     return dt
+
+
+def get_implicit_and_explicit_flat_args(abstracted_axes, *args, **kwargs):
+    """Get implicit arguments from explicit arguments and abstracted_axes."""
+    axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)
+    explicit_args, _ = tree_flatten(args)
+    in_type = infer_lambda_input_type(axes_specs, explicit_args)
+    implicit_args = _extract_implicit_args(in_type, explicit_args)
+    args_flat = [*implicit_args, *explicit_args]
+    return args_flat
+
+
+def jaxpr_remove_implicit(
+    closed_jaxpr: ClosedJaxpr, out_type: OutputType
+) -> tuple[ClosedJaxpr, OutputType]:
+    """Remove all the implicit result values of the ``closed_jaxpr``."""
+    # Note: a more idiomatic way of doing this would be to re-trace the jaxpr and drop the unneeded
+    # tracers.
+    jaxpr = closed_jaxpr.jaxpr
+    out_keep = list(tuple(zip(*out_type))[1]) if len(out_type) > 0 else []
+    outvars = [o for o, keep in zip(jaxpr._outvars, out_keep) if keep]
+    out_type2 = [o for o, keep in zip(out_type, out_keep) if keep]
+    jaxpr2 = Jaxpr(
+        jaxpr.constvars, jaxpr.invars, outvars, jaxpr.eqns, jaxpr.effects, jaxpr.debug_info
+    )
+    return ClosedJaxpr(jaxpr2, closed_jaxpr.consts), out_type2
+
+
+# fmt:off
+# TODO: remove this patch when https://github.com/google/jax/pull/18579 is merged
+def get_aval2(x):
+    """An extended version of `jax.core.get_aval` which also accepts AbstractValues."""
+    if isinstance(x, AbstractValue):
+        return x
+    elif isinstance(x, Tracer):
+        return x.aval
+    else:
+        return concrete_aval(x)
+
+import jax._src.interpreters.partial_eval  # pylint: disable=wrong-import-position, ungrouped-imports
+
+jax._src.interpreters.partial_eval.get_aval = get_aval2
+# fmt:on
+
+
+def make_jaxpr2(
```

Comment:
I think this function should implement the same function as `make_jaxpr` without any changes besides https://github.com/google/jax/pull/18505 . Otherwise we just drift more and more between us and JAX.

Reply:
I agree that we need to try re-using the Jax functionality first, but I think that we have good reasons to overwrite this particular function:
1. We need to return `out_type` and `out_tree` as a separate results rather than as one "typed shape" structure because their length might not match
2. We should (as I believe now - must) always use the  `infer_lambda_input_type` rather than a abstract_axes switch to infer the `in_type`, as Jax does. This way we depend on a smaller subset of Jax actually, which is better. 
3. Also this function could be reduced even more down to `deduce_aval` plus a nested fn call - I plan to do this in a later PR.

Reply:
> We need to return out_type and out_tree as separate results rather than as one "typed shape" structure because their lengths might not match now due to the presence of implicit outputs in out_type.

Yes, but it depends on where we remove the implicit outputs. This is me just stating a preference but I believe we should keep the code as close to JAX as possible to avoid drifting semantically. Using the commit https://github.com/PennyLaneAI/catalyst/pull/366/commits/6c383d90971afd4523e397630e6ba8c9a63f3480 as HEAD, it was possible to still compile and execute functions with just the changes in https://github.com/google/jax/pull/18505. Maybe this is an indication that removing the implicit types when computing `out_tree` could be contributed upstream.

> We should (as I believe now - must) always use the infer_lambda_input_type rather than a abstract_axes switch to infer the in_type, as Jax does. This way we depend on a smaller subset of Jax actually, which is better.

I am not sure JAX always uses `infer_lambda_input_type`. JAX only uses `infer_lambda_input_type` twice in their repo. One checks for `abstracted_axes`. The other for `dynamic_shapes` (which still depends on the `abstracted_axes` value. [See here](https://github.com/google/jax/blob/2a8180ce22d1e282cdcdf6e673edebba3e80e3f0/jax/_src/api.py#L2447-L2455) and [here](https://github.com/google/jax/blob/main/jax/_src/pjit.py#L470-L472)

```python
  def abstractify(args, kwargs):
    flat_args, in_tree = tree_flatten((args, kwargs))
    if abstracted_axes is None:
      return map(shaped_abstractify, flat_args), in_tree, [True] * len(flat_args)
    else:
      axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)
      in_type = pe.infer_lambda_input_type(axes_specs, flat_args)
      in_avals, keep_inputs = unzip2(in_type)
      return in_avals, in_tree, keep_inputs
 ```

```python
  axes_specs = _flat_axes_specs(abstracted_axes, *args, **kwargs)

  # ... snip ...

  if config.dynamic_shapes.value:
    in_type = pe.infer_lambda_input_type(axes_specs, explicit_args)
    in_avals = tuple(a for a, e in in_type if e)
```

Reply:
I am going to resolve this conversation. Even though I think we should keep the implementation as close to JAX as possible, this is something that could easily be changed in the future and also shouldn't keep this PR from being merged.